lib/TimeSeries/AdaptiveFilter.pm

```diff
@@ -0,0 +1,210 @@
+package TimeSeries::AdaptiveFilter;
+
+=head1 NAME
+
+TimeSeries::AdaptiveFilter - Adaptive filter for data stream with possible outliers
+
+=cut
+
+our $VERSION = '0.02';
+
+=head1 VERSION
+
+Version 0.02
+
+=head1 STATUS
+
+=begin HTML
+
+<p>
+    <a href="https://travis-ci.org/binary-com/perl-TimeSeries-AdaptiveFilter"><img src="https://travis-ci.org/binary-com/perl-TimeSeries-AdaptiveFilter.svg" /></a>
+</p>
+
+=end HTML
+
+=cut
+
+use strict;
+use warnings;
+
+use List::Util qw(sum max);
+
+use parent qw/Exporter/;
+
+our @EXPORT_OK = qw/filter/;
+
+=head1 SYNOPSYS
+
+  use TimeSeries::AdaptiveFilter qw/filter/;
+
+  # creation with defaults
+  my $filter = filter();
+
+  # create filter with tuned parameters
+  my $filter = filter({
+    floor             => 6
+    cap               => 0.2,
+    lookback_capacity => 20,
+    lookback_period   => 4,
+    decay_speeds      => [0.03, 0.01, 0.003],
+    build_up_count    => 5,
+    reject_criterium  => 4,
+  });
+
+  # usage
+  my $now = time;
+  $filter->($now, 100.002);        # returns true, i.e. all data is valid on learning period
+  $filter->($now + 1, 100.001);    # returns true
+  ...                              # it learns form sample of 60 seconds
+  $filter->($now + 60, 100.005);   # returns true
+  $filter->($now + 61, 99.9995);   # returns true, as value does not differs much
+  $filter->($now + 62, 10_0000);   # returns false, outlier data
+  $filter->($now + 63, 10.0001);   # returns false, outlier data
+  $filter->($now + 64, 100.011);   # returns true, even if the sample is oulier, because
+                                   # the filter rejected too much values, and has to
+                                   # re-adapt to time seria again
+
+=head1 DESCRIPTION
+
+For the details of underlying mathematical model of the filter, configurable paramters
+and their usage, please, look at the shipped C<doc> folder.
+
+=cut
+
+
+# private lexical function, which calculates MAD (mean absolute deviation)
+my $_mad = sub {
```

Comment:
> is there any reason why this has to be a function reference instead of a normal function?

Actually, I would prefer as private as possible, because people in binary like to call private-by-convention (i.e. with `_` prefix) functions directly. 

With lexical scoped things, they have no chance to do that. 

Anyway, I'll inline it as you suggest.